qermit/coherent_pauli_checks/monochromatic_convex_subdag.py

```diff
@@ -0,0 +1,147 @@
+from itertools import permutations
+from typing import Any
+
+import networkx as nx  # type: ignore
+
+
+def _subdag_nodes(subdag: int, node_subdag: dict[Any, int]) -> list[Any]:
+    """Get all nodes in a given sub-DAG.
+
+    :param subdag: The sub-DAG whose nodes should be retrieved.
+    :param node_subdag: Map from node to the sub-DAG to which the
+        node belongs.
+    :return: List of nodes belonging to given sub-DAG.
+    """
+    return [node for node, s in node_subdag.items() if s == subdag]
+
+
+def _subdag_predecessors(
+    dag: nx.DiGraph, subdag: int, node_subdag: dict[Any, int]
+) -> list[Any]:
+    """Retrieve all nodes not in given sub-DAG with successors in sub-DAG.
+
+    :param dag: Directed Acyclic Graph.
+    :param subdag: Sub-DAG to retrieve predecessors of.
+    :param node_subdag: Map from node to the sub-DAG to which it belongs.
+    :return: Nodes with successors in given sub-DAG.
+    """
+    subdag_nodes = _subdag_nodes(subdag=subdag, node_subdag=node_subdag)
+    return sum(
+        [
+            [
+                predecessor
+                for predecessor in dag.predecessors(node)
+                # Exclude nodes in subdag.
+                if predecessor not in subdag_nodes
+            ]
+            for node in subdag_nodes
+        ],
+        start=[],
+    )
+
+
+def _subdag_successors(
+    dag: nx.DiGraph, subdag: int, node_subdag: dict[Any, int]
+) -> list[Any]:
+    """Retrieve all nodes not in given sub-DAG with predecessors in sub-DAG.
+
+    :param dag: Directed Acyclic Graph.
+    :param subdag: Sub-DAG to retrieve successors of.
+    :param node_subdag: Map from node to the sub-DAG to which it belongs.
+    :return: Nodes with predecessors in given sub-DAG.
+    """
+    subdag_nodes = _subdag_nodes(subdag=subdag, node_subdag=node_subdag)
+    return sum(
+        [
+            [
+                successor
+                for successor in dag.successors(node)
+                # Exclude nodes in subdag.
+                if successor not in subdag_nodes
+            ]
+            for node in subdag_nodes
+        ],
+        start=[],
+    )
+
+
+def get_monochromatic_convex_subdag(
+    dag: nx.DiGraph, coloured_nodes: set[Any]
+) -> dict[Any, int]:
+    """Retrieve assignment of coloured nodes to sub-DAGs.
+    The assignment aims to minimise the number of sub-DAGs.
+
+    :param dag: Directed Acyclic Graph.
+    :param coloured_nodes: The nodes which are coloured.
+    :return: Map from node to the sub-DAG to which it belongs.
+    """
+    node_subdag = {node: i for i, node in enumerate(coloured_nodes)}
+
+    node_descendants = {node: nx.descendants(dag, node) for node in dag.nodes}
+    for node in node_descendants.keys():
+        node_descendants[node].add(node)
+
+    def _can_merge(
+        dag: nx.DiGraph,
+        subdag_one: int,
+        subdag_two: int,
+        node_subdag: dict[Any, int],
```

Comment:
It's rather confusing to have `node_subdag` as an argument to this function when it is also a variable in the surrounding scope. Does it even need to be passed as an argument? I would probably implement this as a class method instead.

Reply:
It does need to be an input as `node_subdag` changes between calls. The thing that I want to fix and calculate only once is `node_descendants`.
For the reasons discussed below I don't feel that this should be a class, but I could probably be persuaded either way.